frontend/src/Frontend/KadenaAddress.hs

```diff
@@ -0,0 +1,243 @@
+import qualified Data.Text.Encoding.Error as TE
+
+import qualified Data.Attoparsec.ByteString.Char8 as A
+
+import qualified URI.ByteString as URI
+import qualified Data.ByteString.Base64 as Base64
+import qualified Data.Digest.CRC32 as Digest
+
+import Pact.Types.ChainId
+import Kadena.SigningApi (AccountName(..), mkAccountName)
+
+import Common.Network (NetworkName,textNetworkName,mkNetworkName)
+
+{-
+Encoder / Decoder should not be derived automatically as package updates could silently
+change the format and break things. Keep to simple hand-rolled parser & printer that
+satisfies the round-trip property: (print . parse = id), hedgehog has 'tripping' for this.
+
+Final encoding will be Base64 because it needs to be reliably cut'n'paste between
+platforms without risking corruption.
+
+The address will contain the following fields:
+
+AccountName: Compliant with the requirements contained in pact. In the case of wallet-only
+accounts that do not exist on the blockchain, the Public Key will be used.
+ChainId: Non-negative integer.
+Network: The network where the chain resides.
+Checksum: The checksum will be CRC32 of the 'AccountName/PublicKey', 'ChainId',
+‘Network’. An invalid checksum will be a parse failure.
+
+Some examples of the encoding (delimiter yet to be determined):
+
+Using newlines:
+doug
+1
+us1.tn1.chainweb.com
+1bfafb513a26979b48bafe3f9c0fba7703065066a84421f71ca41b54692de67d
+71f920fa275127a7b60fa4d4d41432a3
+
+Or some other delimiter:
+doug%1%us1.tn1.chainweb.com%1bfafb513a26979b48bafe3f9c0fba7703065066a84421f71ca41b54692de67d%71f920fa275127a7b60fa4d4d41432a3
+
+The final form includes a human readable component as well as machine readable components:
+
+<account name>][<chain id>][<base64 encoding of address>
+
+The use of an address does not require the human readable component. Only the Base64
+encoded address is validated against the checksum. If the human readable prefix is
+included then it will be checked against the values decoded in the address as an extra
+preventative measure.
+-}
+
+data KadenaAddressError
+  = ParseError Text
+  | CRC32Mismatch Word32 Word32
+  | Base64Error String
+  | InvalidHumanReadablePiece Text
+  deriving (Show, Eq)
+
+data KadenaAddress = KadenaAddress
+  { _kadenaAddress_encoded :: ByteString
+    -- ^ The complete encoded version of this address
+  , _kadenaAddress_accountName :: AccountName
+    -- ^ The account name associated with this address
+  , _kadenaAddress_chainId :: ChainId
+    -- ^ The chain where this account resides
+  , _kadenaAddress_networkName :: NetworkName
+    -- ^ The network where this chain resides
+  , _kadenaAddress_checksum :: Word32
+    -- ^ The CRC32 checksum of the AccountName, ChainId, and NetworkName
+  }
+  deriving (Show, Eq)
+
+textKadenaAddress :: KadenaAddress -> Text
+textKadenaAddress = TE.decodeUtf8With TE.lenientDecode . _kadenaAddress_encoded
+
+humanReadableDelimiter :: Word8
+humanReadableDelimiter = 93 -- ']'
```

Comment:
I'd have a minor preference for `|`. No reason other than it looks better :smile:

Reply:
Come to think of it, account names can have this character.

Reply:
Account names can currently have everything in latin1, so this was a placeholder at best. :< Also I agree that `|` looks better. I'll make a ticket.